File: C++/programas/P-Calificaciones alumnos/P-Calificaciones alumnos/CaliAl.cpp
#include "CaliAl.h" //Cabecera de clase CaliAl.
#include <iostream> //Cabecera para entrada y salida de datos.

//M?todo que ingresa elementos a los arreglos.
void CaliAl::IngresoDatos(float cal[], std::string nom[], int cont)
{
	int i;

	if (cont == 5) //Condici?n para salir del m?todo recursivo.
	{
		Prom = AcumCali / 5; //Asignaci?n de valor a variable.
		for (i = 0; i < 5; i++) //Clasificaci?n de datos ingresados.
		{
			if (cal[i] > CaliMax)
				CaliMax = cal[i]; //Asignaci?n de valor a variable.
		}

		CaliMen = CaliMax; //Asignaci?n de valor a variable.
		for (i = 0; i < 5; i++)
		{
			if (cal[i] < CaliMen)
				CaliMen = cal[i]; //Asignaci?n de valor a variable.
		}

		Despliegue(cal, nom); //Llamada a m?todo que despliega elementos del arreglo y datos caracter?sticos.
	}
	else
	{
		//Ingreso de dato a variable.
		std::cout << (cont + 1) << ". Ingresa una calificaci" << char(162) << "n: "; std::cin >> cal[cont];
		if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
		{
			std::cout << "Error de formato.\n";
			system("pause");
			std::cin.clear(); //Limpia el buffer de cin.
			//Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
			std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
			IngresoDatos(cal, nom, cont); //Llamada recursiva a IngresoDatos.
		}
		else
		{
			//Ingreso de dato a variable.
			std::cout << (cont + 1) << ". Ingresa un nombre: "; std::cin >> nom[cont];

			//Clasificaci?n de datos ingresados.
			AcumCali += cal[cont];
			if (cal[cont] > 0.0 && cal[cont] <= 10.0)
			{
				CalVali++;
				if (cal[cont] > 6.0 && cal[cont] <= 10.0)
					CaliApro++;
				else if (cal[cont] <= 6.0 && cal[cont] > 0.0)
					CaliRepro++;
			}
			else
				CalInvali++;

			cont++; //Incrementa contador.
			IngresoDatos(cal, nom, cont); //Llamada recursiva a IngresoDatos.
		}
	}
}

//M?todo que despliega los elementos de los arreglos.
void CaliAl::Despliegue(float cal[], std::string nom[])
{
	int i;

	//Despliegue de elementos del arreglo.
	std::cout << "\nLos alumnos ingresados son:\n\n";
	for (i = 0; i < 5; i++)
	{
		std::cout << (i + 1) << " - " << nom[i] << ", con calificaci" << char(162) << "n de " << cal[i] << ".\n";
	}

	//Despliegue de datos caracter?sticos del arreglo.
	std::cout << "Hubo un total de " << CalVali << " calificaciones v" << char(160) << "lidas y " << CalInvali << " inv" << char(160) << "lidas.\n";
	std::cout << "Hubo un total de " << CaliApro << " calificaciones aprobatorias y " << CaliRepro << " reprobatorias.\n";
	std::cout << "La calificaci" << char(162) << "n mayor fue " << CaliMax << " y la menor es " << CaliMen << ".\n";
	std::cout << "El promedio es de " << Prom << ".";
}

File: Java/programas/CalificacionesAlumnos/src/CaliAl.java
import java.util.Scanner; //Librería para declara objeto Scanner de entrada y salida.
import java.util.InputMismatchException; //Excepción si hay un error de formato en tipos de datos.

public class CaliAl {
    public byte CalVali = 0, CalInvali = 0, CaliApro = 0, CaliRepro = 0;
    public float Prom, AcumCali = 0, CaliMax = 0, CaliMen; //Atributos públicos.
    Scanner lectura = new Scanner(System.in); //Declara objeto Scanner.

    //Procedimiento que recibe calificaciones para luego clasificarlas según:
    //Calificaciones válidas (>=0 y <=10), inválidas (<0 y >10), aprobatorias (>5),
    //reprobatorias (<=5), calificación mayor, menor y suma de todas las calificaciones.
    public void IngresoDatos(float[] cal, String[] nom, int cont){
        int i;

        if (cont == 5){ //Condición para salir del método recursivo.
            Prom = AcumCali / 5; //Calcula promedio.
            for (i = 0; i < 5; i++){ //Obtiene calificación mayor del arreglo de calificaciones.
                if (cal[i] > CaliMax){
                    CaliMax = cal[i];
                }
            }
    
            CaliMen = CaliMax; //Establece como referencia la calificación mayor para encontrar la menor.
            for (i = 0; i < 5; i++){ //Obtiene calificación menor del arreglo de calificaciones.
                if (cal[i] < CaliMen){
                    CaliMen = cal[i];
                }
            }

            lectura.close(); //Cierra el objeto Scanner.
            Despliegue(cal, nom); //Llamada al método que despliega resultados
        } else{
            //Selectiva try-catch por si se ingresa un tipo de dato incorrecto en la variable.
            try{
                System.out.printf("%s. Ingresa una calificación: ", (cont + 1)); cal[cont] = lectura.nextFloat();
                lectura.nextLine();
                System.out.printf("%s. Ingresa un nombre: ", (cont + 1)); nom[cont] = lectura.nextLine();

                AcumCali += cal[cont]; //Suma todas las calificaciones.
                if (cal[cont] > 0.0 && cal[cont] <= 10.0){ //Clasifica calificación en válida o inválida.
                    CalVali++;
                    if (cal[cont] > 6.0 && cal[cont] <= 10.0){ //Clasifica calificación en aprobatoria o reprobatoria.
                        CaliApro++;
                    } else if (cal[cont] <= 6.0 && cal[cont] > 0.0){
                        CaliRepro++;
                    }  
                } else{
                    CalInvali++;
                }

                cont++; //Incrementa contador.
                IngresoDatos(cal, nom, cont); //Llamada recursiva.
            } catch(InputMismatchException e){
                System.out.println("Error de formato.");
                lectura.nextLine(); //Limpia buffer de entrada de datos.
                IngresoDatos(cal, nom, cont);
            }
        }
    }

    //Procedimiento que despliega valores de los atributos.
    public void Despliegue(float[] cal, String[] nom){
        int i;

        System.out.println("\nLos alumnos ingresados son:\n");
        for(i = 0; i < 5; i++){ //Despliega todas las calificaciones.
            System.out.printf("%s - %s, con calificación de %s.\n", (i + 1), nom[i], cal[i]);
        }

        //Despliega valores de los atributos.
        System.out.printf("Hubo un total de %s calificaciones válidas y %s inválidas.\n", CalVali, CalInvali);
        System.out.printf("Hubo un total de %s calificaciones aprobatorias y %s reprobatorias.\n", CaliApro, CaliRepro);
        System.out.printf("La calificación mayor fue de %s y la menor es de %s.\n", CaliMax, CaliMen);
        System.out.printf("El promedio de todas las calificaciones es de %s.", Prom);
    }
}
